src/ecs/entity.lisp

```diff
@@ -1,5 +1,107 @@
-(in-package :lbge.ecs)
+(in-package :lbge.ecs.entity-storage)
 
-(defclass entity ()
-  ((uuid)
-   (world)))
+(defclass entity-storage ()
+  ((sparse
+    :initform (make-array '(0)
+                          :adjustable t
+                          :element-type '(unsigned-byte 64))
+    :documentation "Sparse set of the entity storage")
+   (packed
+    :initform (make-array '(0)
+                          :adjustable t
+                          :fill-pointer 0
+                          :element-type '(unsigned-byte 64))
+    :documentation "Packed array of entities")
+   (available
+    :initform 0
+    :documentation "Number of entities available for recycling")
+   (create-next
+    :initform 1
+    :documentation "Next entity to be created")
+   (recycle-next
+    :initform (null-entity)
+    :documentation "Next entity to be recycled")))
+
+(defmacro null-entity ()
+  0)
+
+(defmacro is-null (entity)
+  (= entity (null-entity)))
+
+(defmethod initialize-instance :after ((storage entity-storage) &key)
+  :documentation "Add default null entity"
+  (with-slots (sparse packed) storage
+    (ss:insert (null-entity) sparse packed)))
+
+(defun delete-entity (storage entity)
+  "Delete the entity.
+
+The algorithm:
+1. Increase the gen part of the entity
+2. Swap recycle-next with the entity in the packed array
+3. Increase available counter"
+  (with-slots (sparse packed
+               available recycle-next)
+      storage
+    (unless (ss:existsp entity sparse packed)
+      ;; Already deleted
+      (return-from delete-entity nil))
+    (let ((new-recycle-next entity))
+      (setf
+       ;; Increase gen
+       (ss:gen new-recycle-next)
+       (1+ (ss:gen entity))
+
+       ;; Set recycle-next to the current one
+       (aref packed
+             (aref sparse (ss:id entity)))
+       recycle-next
+
+       ;; Current one will be recycled next
+       recycle-next
+       new-recycle-next
+
+       ;; Increase available counter
+       available
+       (1+ available)))))
+
+(defun create-entity (storage)
+  "Create entity (or recycle one created earlier).
+
+The recycle algorithm:
+1. Restore previous recycle-next entity
+2. Decrease available count
+3. Restore original recycle-next as live and return it"
+  (with-slots (create-next
+               recycle-next
+               available
+               sparse packed)
+      storage
+    (when (= 0 available)
```

Comment:
Wouldn't it be better to use `(zerop available)` instead of a direct comparison to a literal zero?

Reply:
Yes, it indeed will be better.
I forget about specific predicates all the time.